#622 Init example for project template Java/EMF Model/Theia application

ndoschek opened the pull request:
- Setup glsp server and client
- Init EMF model resource loading
- Consume latest glsp server and client and latest example models from modelserver-glsp-integration
- Supports creation and deletion of nodes
- Includes the generic EMF classes, this PR needs to be updated after those are contributed to the glsp server

Part of https://github.com/eclipse-glsp/glsp/issues/622

Contributed on behalf of STMicroelectronics.

planger commented on project-templates/java-emf-theia/glsp-server/src/main/java/org/eclipse/glsp/example/javaemftheia/emf/EMFResourceManager.java:
```diff
@@ -0,0 +1,49 @@
+/**
+ * This resource manager allows to load, save and manipulate EMF resources.
+ */
+public interface EMFResourceManager {
```
As mentioned above, I'm not quite sure we should introduce another concept here.
This class combines
1. the SourceModelStorage functionality which could for EMF operate on ResourceSet level
2. the CommandStack, which could just be specialized (if necessary) and used by the gmodel state and all others as they do by default

ndoschek replied:
Yes I agree, the resource manager was removed with the follow-up commit.

planger commented on project-templates/java-emf-theia/glsp-server/src/main/java/org/eclipse/glsp/example/javaemftheia/emf/EMFResourceManagerImpl.java:
```diff
@@ -0,0 +1,193 @@
+import org.eclipse.emf.edit.domain.AdapterFactoryEditingDomain;
+import org.eclipse.glsp.server.types.GLSPServerException;
+
+public abstract class EMFResourceManagerImpl implements EMFResourceManager {
```
See comment in `EMFResourceManager` interface.

planger commented on project-templates/java-emf-theia/glsp-server/src/main/java/org/eclipse/glsp/example/javaemftheia/emf/EMFSourceModelStorage.java:
```diff
@@ -0,0 +1,59 @@
+ * This implementation of the {@link SourceModelStorage} handles the persistence of source models via
+ * the {@link EMFResourceManager}.
+ */
+public abstract class EMFSourceModelStorage implements SourceModelStorage {
```
This should be moved into the `glsp-server` eventually.

planger commented on project-templates/java-emf-theia/glsp-server/src/main/java/org/eclipse/glsp/example/javaemftheia/launch/TaskListCLIParser.java:
```diff
@@ -0,0 +1,39 @@
+import org.apache.commons.cli.ParseException;
+import org.eclipse.glsp.server.launch.DefaultCLIParser;
+
+public class TaskListCLIParser extends DefaultCLIParser {
```
Do we need this class for the Theia case?

planger commented on project-templates/java-emf-theia/glsp-server/src/main/java/org/eclipse/glsp/example/javaemftheia/model/TaskListResourceManager.java:
```diff
@@ -0,0 +1,95 @@
+import org.eclipse.glsp.server.types.GLSPServerException;
+
+@Singleton
+public class TaskListResourceManager extends EMFResourceManagerImpl {
```
I feel this all would better fit into the task model source loader

ndoschek commented on project-templates/java-emf-theia/glsp-server/src/main/java/org/eclipse/glsp/example/javaemftheia/emf/EMFDiagramFactory.java:
```diff
@@ -0,0 +1,86 @@
+ * Subclasses need to implement `EMFDiagramFactory#addDiagramChildren` and translate the semantic
+ * and notation model information into GModel elements.
+ */
+public abstract class EMFDiagramFactory {
```
This was removed with the follow-up commit

ndoschek commented on project-templates/java-emf-theia/glsp-client/tasklist-browser-app/package.json:
```diff
@@ -0,0 +1,32 @@
+    "@theia/process": "1.24.0",
+    "@theia/terminal": "1.24.0",
+    "@theia/workspace": "1.24.0",
+    "@eclipse-glsp-examples/tasklist-theia": "0.10.0"
```
I'm afraid I do not really get your comment, `@eclipse-glsp-examples/tasklist-theia` is the theia integration of this example I'm working on.

ndoschek commented on project-templates/java-emf-theia/glsp-client/tasklist-theia/package.json:
```diff
@@ -0,0 +1,52 @@
+    "src"
+  ],
+  "dependencies": {
+    "@eclipse-glsp-examples/tasklist-glsp": "0.10.0",
```
Same here

ndoschek commented on project-templates/java-emf-theia/glsp-client/tasklist-theia/src/browser/diagram/tasklist-diagram-configuration.ts:
```diff
@@ -0,0 +1,31 @@
+    diagramType: string = TaskListLanguage.diagramType;
+
+    doCreateContainer(widgetId: string): Container {
+        const container = createTaskListDiagramContainer(widgetId);
```
Same here

ndoschek commented on project-templates/java-emf-theia/glsp-server/src/main/java/org/eclipse/glsp/example/javaemftheia/emf/EMFGModelFactory.java:
```diff
@@ -0,0 +1,206 @@
+    public void createGModel() {
+        EMFGModelState modelState = EMFGModelState.getModelState(gModelState);
+        EObject taskList = modelState.getSemanticModel();
+        Diagram diagram = modelState.getNotationModel();
```
Yes I agree basically, but I kept it in my follow-up commit, as it is needed to initialize the GModelIndex properly.
Could you have another look what you think of it now, or if you think the index initialization should be moved to the specific TaskList implementation perhaps?

ndoschek commented on project-templates/java-emf-theia/glsp-server/src/main/java/org/eclipse/glsp/example/javaemftheia/emf/EMFGModelState.java:
```diff
@@ -0,0 +1,109 @@
+public abstract class EMFGModelState extends DefaultGModelState implements ClientSessionListener {
+
+    @Inject
+    protected EMFResourceManager resourceManager;
```
For now I reused the `commandStack` memeber of the `DefaultGModelState`

ndoschek commented on project-templates/java-emf-theia/glsp-server/src/main/java/org/eclipse/glsp/example/javaemftheia/emf/EMFGModelState.java:
```diff
@@ -0,0 +1,109 @@
+        this.clientSessionManager.addListener(this, this.clientId);
+    }
+
+    public abstract EObject getSemanticModel();
```
https://github.com/eclipse-glsp/glsp-examples/pull/109#discussion_r856124260

ndoschek commented:
Thanks a lot for your review @planger, I removed and moved a lot of responsibilities as you suggested. Could you please have another look and if you see further improvements? Thanks again!

planger commented on project-templates/java-emf-theia/glsp-server/src/main/java/org/eclipse/glsp/example/javaemf/TaskListDiagramConfiguration.java:
```diff
@@ -0,0 +1,46 @@
+
+   @Override
+   public List<EdgeTypeHint> getEdgeTypeHints() {
+      EdgeTypeHint transitionEdgeHint = new EdgeTypeHint(TaskListModelTypes.TRANSITION, false, false, false,
```
I think it would be good if the edge would at least be deletable.

ndoschek replied:
I removed the edge type hint for now as we do not support it yet, I will probably re-add it once the remaining implementation is stable enough.

planger commented on project-templates/java-emf-theia/glsp-server/src/main/java/org/eclipse/glsp/example/javaemf/TaskListDiagramConfiguration.java:
```diff
@@ -0,0 +1,46 @@
+
+   @Override
+   public List<ShapeTypeHint> getShapeTypeHints() {
+      return List.of(new ShapeTypeHint(TaskListModelTypes.TASK, false, true, false, false));
```
I think it would be great if the node would be movable and resizable.

ndoschek replied:
Yes I agree, I added that with the current commit

planger commented on project-templates/java-emf-theia/glsp-server/src/main/java/org/eclipse/glsp/example/javaemf/emf/EMFGModelFactory.java:
```diff
@@ -0,0 +1,98 @@
+public abstract class EMFGModelFactory implements GModelFactory {
+
+    @Inject
+    protected GModelState gModelState;
```
I think you could inject the type `EMFGModelState` here directly and avoid having to cast later on.

ndoschek replied:
I injected the specific ModelState type where possible (i.e. the `TaskListGModelState`). For the others (`EMFGModelState` and `EMFNotationGModelState`) I went with the casting approach for very few occurrences as we would have different instances (as we only bind `TaskListGModelState` directly) which causes problems.

planger commented on project-templates/java-emf-theia/glsp-server/src/main/java/org/eclipse/glsp/example/javaemf/emf/EMFGModelState.java:
```diff
@@ -0,0 +1,175 @@
+
+        URI resourceURI = URI.createFileURI(sourceURI);
+        Optional<Resource> loadedResource = Optional.ofNullable(this.resourceSet.getResource(resourceURI, false))
+            .filter(Resource::isLoaded);
+
+        Resource resource = null;
+        if (loadedResource.isPresent()) {
+            resource = loadedResource.get();
+        } else {
+            resource = this.resourceSet.getResource(resourceURI, true);
+            resource.load(Collections.EMPTY_MAP);
+        }
+
+        EcoreUtil.resolveAll(resource);
+        EObject root = resource.getContents().get(0);
+        if (!modelClass.isInstance(root)) {
+            LOGGER.info("Resource appears to be empty: " + sourceURI);
+            return null;
+        }
+        return modelClass.cast(root);
+    }
+
+    public void saveResources() {
+        boolean result = false;
+        EList<Resource> resourceList = this.resourceSet.getResources();
+        for (Resource resource : resourceList) {
+            if (resource.getURI() != null) {
+                try {
+                    resource.save(Collections.EMPTY_MAP);
+                    result = true;
+                } catch (IOException e) {
+                    LOGGER.error("Could not save resource: " + resource.getURI(), e);
+                    throw new GLSPServerException("Could not save model to file: " + resource.getURI(), e);
+                }
+            }
+        }
+        if (result) {
+            this.commandStack.saveIsDone();
+        }
+    }
```
I would put this code into an `EMFSourceModelStorage`.

planger replied:
It doesn't show nicely anymore in Github. I meant here the saveResource and loadResource methods.
In general my goal would be that the model state is not concerned with loading or saving at all, but just holds the resource set (and closes it on dispose).

The source model loader should be responsible for loading and saving.
Also it is responsible to parse which URI to load from the arguments, etc.

ndoschek replied:
Ah thanks, I overlooked that in the first run I guess. With the current commit this should be resolved and the `SourceModelStorage` is responsible for the resource handling. The ResourceSet is stored in the ModelState.

planger commented on project-templates/java-emf-theia/glsp-server/src/main/java/org/eclipse/glsp/example/javaemf/emf/EMFGModelState.java:
```diff
@@ -0,0 +1,175 @@
+            }
+        }
+        if (result) {
+            this.commandStack.saveIsDone();
```
I think this is done by the generic `SaveModelActionHandler`.

planger commented on project-templates/java-emf-theia/glsp-server/src/main/java/org/eclipse/glsp/example/javaemf/emf/EMFGModelState.java:
```diff
@@ -0,0 +1,175 @@
+        EObject root = resource.getContents().get(0);
+        if (!modelClass.isInstance(root)) {
+            LOGGER.info("Resource appears to be empty: " + sourceURI);
+            return null;
```
Please rather return optional here.

planger commented on project-templates/java-emf-theia/glsp-server/src/main/java/org/eclipse/glsp/example/javaemf/emf/EMFGModelState.java:
```diff
@@ -0,0 +1,175 @@
+
+    public AdapterFactoryEditingDomain getEditingDomain() { return editingDomain; }
+
+    protected abstract void setupResourceSet();
+
+    public abstract void loadResources();
```
I'd would see those in the responsibility of an `EMFSourceModelStorage`.

planger commented on project-templates/java-emf-theia/glsp-server/src/main/java/org/eclipse/glsp/example/javaemf/emf/EMFGModelState.java:
```diff
@@ -0,0 +1,175 @@
+    @Override
+    public EMFModelIndex getIndex() { return EMFModelIndex.get(getRoot()); }
+
+    public static EMFGModelState getModelState(final GModelState state) {
+        if (!(state instanceof EMFGModelState)) {
+            throw new IllegalArgumentException("Argument must be a EMFGModelState");
+        }
+        return ((EMFGModelState) state);
+    }
```
I'm not sure whether this is ever needed. I think every other class can get the concrete subtype injected.

planger commented on project-templates/java-emf-theia/glsp-server/src/main/java/org/eclipse/glsp/example/javaemf/emf/EMFGModelState.java:
```diff
@@ -0,0 +1,175 @@
+    protected ResourceSet resourceSet;
+    protected AdapterFactoryEditingDomain editingDomain;
+
+    public abstract EObject getSemanticModel();
+
+    public abstract Diagram getNotationModel();
```
I'd prefer if these two methods wouldn't be declared here, but rather be only implemented by subclasses.
Others who need these methods could get the subclass injected, if they need.
But this way this model state class is completely agnostic of how the source model(s) is/are structured.

ndoschek replied:
Agreed, moved that to the `TaskListGModelState` with the current commit.

planger commented on project-templates/java-emf-theia/glsp-server/src/main/java/org/eclipse/glsp/example/javaemf/emf/EMFModelIndex.java:
```diff
@@ -0,0 +1,271 @@
+ * Is used to index all child elements of a semantic and notation model.
+ * Offers a set of query methods retrieve indexed elements.
+ */
+public final class EMFModelIndex extends GModelIndexImpl {
```
I think it'd be great if there would be two index implementations:
1. The `EMFModelIndex` which maps GModel elements to `EObjects` (this one doesn't make any assumption on whether it is a notation or semantic element and can also be used if there is no separation.
2. A `EMFNotationModelIndex` that maps Semantic objects to Notation objects. This could be an extension of the `EMFModelIndex`

ndoschek replied:
Done with the current commit

planger commented on project-templates/java-emf-theia/glsp-server/src/main/java/org/eclipse/glsp/example/javaemf/emf/EMFSourceModelStorage.java:
```diff
@@ -0,0 +1,46 @@
+    protected GModelState gModelState;
+
+    @Override
+    public void loadSourceModel(RequestModelAction action) {
```
Please see my comments in the EMFGModelState.

planger commented on project-templates/java-emf-theia/glsp-server/src/main/java/org/eclipse/glsp/example/javaemf/emf/EMFGModelFactory.java:
```diff
@@ -0,0 +1,98 @@
+ * It expects a semantic as well as an notation model to create to the GModel.
+ * Subclasses need to provide a diagram factory that translates the semantic and notation model information to a GModel.
+ */
+public abstract class EMFGModelFactory implements GModelFactory {
```
As far as I see, this is really only adding a lot of value for the notation/semantic model use case, right?
I wonder if we just rename that to `EMFNotationModelFactory` and avoid binding/enforcing it by default in the `EMFDiagramModule`.
Or we just push this down to the concrete Task model factory altogether and skip it as a generic thing that we offer.

ndoschek replied:
Yes I agree, I removed it from the base implementation and implemented the `TaskListModelFactory`

planger reviewed (commented):
Hi Nina, thanks a lot! Here another round of comments (have a meeting now, but didn't want to wait to share those). Maybe there are a few more, but I think the most important ones are captured.

I think we are on a great track here! Thanks for your work!

ndoschek commented:
Thanks again @planger for the review. I reworked the responsibilities of `SourceModelStorage`, `GModelState` and the `GModelFactory`.
Also, I splitted the base classes in base-base and notation-base classes and put them in a separate folder for now (in preparation for the GLSP server contribution).
And I added the implementation for moving/resizing nodes.
Edge support is currently not covered, but if there is some time left I will add that as last point once the base classes are extracted and the example basis is stable.

planger reviewed (changes requested):
Looks great, thanks! A few inline suggestions.
Also the java build failed for me:
```
[INFO] -------------------------------------------------------------
[ERROR] COMPILATION ERROR : 
[INFO] -------------------------------------------------------------
[ERROR] /home/philip/Git/OpenSource/eclipse-glsp/glsp-examples/project-templates/java-emf-theia/glsp-server/src/main/java/org/eclipse/glsp/example/javaemf/model/TaskListSourceModelStorage.java:[19,47] package org.eclipse.glsp.example.tasklist.model does not exist
[ERROR] /home/philip/Git/OpenSource/eclipse-glsp/glsp-examples/project-templates/java-emf-theia/glsp-server/src/main/java/org/eclipse/glsp/example/javaemf/handler/CreateTaskNodeHandler.java:[29,47] package org.eclipse.glsp.example.tasklist.model does not exist
[ERROR] /home/philip/Git/OpenSource/eclipse-glsp/glsp-examples/project-templates/java-emf-theia/glsp-server/src/main/java/org/eclipse/glsp/example/javaemf/handler/CreateTaskNodeHandler.java:[30,47] package org.eclipse.glsp.example.tasklist.model does not exist
[ERROR] /home/philip/Git/OpenSource/eclipse-glsp/glsp-examples/project-templates/java-emf-theia/glsp-server/src/main/java/org/eclipse/glsp/example/javaemf/handler/CreateTaskNodeHandler.java:[31,47] package org.eclipse.glsp.example.tasklist.model does not exist
[ERROR] /home/philip/Git/OpenSource/eclipse-glsp/glsp-examples/project-templates/java-emf-theia/glsp-server/src/main/java/org/eclipse/glsp/example/javaemf/handler/CreateTaskNodeHandler.java:[32,47] package org.eclipse.glsp.example.tasklist.model does not exist
[ERROR] /home/philip/Git/OpenSource/eclipse-glsp/glsp-examples/project-templates/java-emf-theia/glsp-server/src/main/java/org/eclipse/glsp/example/javaemf/handler/CreateTaskNodeHandler.java:[94,14] cannot find symbol
  symbol:   class Task
  location: class org.eclipse.glsp.example.javaemf.handler.CreateTaskNodeHandler
...
```

ndoschek commented on project-templates/java-emf-theia/glsp-server/src/main/java/org/eclipse/glsp/example/javaemf/model/TaskListGModelState.java:
```diff
@@ -0,0 +1,56 @@
+    public void setNotationModel(final Diagram diagram) { this.notationModel = diagram; }
+
+    @Override
+    protected void setupResourceSet() {
```
It lives already in the storage, this file does not exist anymore, unfortunately you had an outdated state I assume.

ndoschek commented on project-templates/java-emf-theia/glsp-server/src/main/java/org/eclipse/glsp/example/javaemf/model/TaskListGModelState.java:
```diff
@@ -0,0 +1,56 @@
+
+    public Diagram getNotationModel() { return this.notationModel; }
+
+    public void setNotationModel(final Diagram diagram) { this.notationModel = diagram; }
```
Same here

ndoschek commented on project-templates/java-emf-theia/glsp-server/src/main/java/org/eclipse/glsp/example/javaemf/model/TaskListGModelState.java:
```diff
@@ -0,0 +1,56 @@
+
+    public TaskList getSemanticModel() { return this.semanticModel; }
+
+    public void setSemanticModel(final TaskList taskList) { this.semanticModel = taskList; }
```
Same here

ndoschek commented on project-templates/java-emf-theia/glsp-server/src/main/java/org/eclipse/glsp/example/javaemf/handler/DeleteTaskNodeHandler.java:
```diff
@@ -0,0 +1,76 @@
+            return Optional.empty();
+        }
+        EMFNotationModelIndex index = modelState.getIndex();
+        Optional<GModelElement> element = index.get(elementIds.get(0));
```
Same here, this is an outdated version, we delete all selected elements now

ndoschek commented on project-templates/java-emf-theia/glsp-server/src/main/java/org/eclipse/glsp/example/javaemf/emf/notation/EMFNotationModelIndex.java:
```diff
@@ -0,0 +1,236 @@
+ * Is used to index all child elements of a semantic and notation model.
+ * Offers a set of query methods retrieve, initialize and resolve indexed notoation elements.
+ */
+public final class EMFNotationModelIndex extends EMFModelIndex {
```
Same here

ndoschek commented:
Thanks @planger for your review, I updated the `pom.xml` there was something off, now it should build consistently without errors.
I updated the README as well, regarding the remaining source code comments, they seem to be fixed with the last revision.

planger reviewed (approved):
Works now, thanks!
Sorry, I think there were some old pending review comments included in my last review, that have accidently been published with my latest review.
I double-checked, the new comments have all been incorporated into the Readme and now building and running works fine.

I just observed a bug, but this is likely a general bug rather than specific to this example, so I created a GH issue for it: https://github.com/eclipse-glsp/glsp/issues/635
It shouldn't block merging this PR.
